### DISCUSSION TITLE
Trouble Integrating Custom LCD Driver as a Zephyr Module

### DISCUSSION THREAD

**Discussion by saacifuentesmu-titoma** [zephyrproject-rtos/zephyr#61914] [Q&A] [answered]
I work for a B2B company that manufactures devices in China. One of our devices currently runs a bare-metal program, and I successfully reimplemented its functionality using Zephyr. Initially, I was quite enthusiastic about the ease of development and was keen to advocate for the adoption of Zephyr within my organization, primarily due to its portability and extensive driver support.

However, I've encountered a roadblock that has made it difficult to proceed. I have a device featuring a custom LCD, and my goal is to use LVGL on Zephyr for its UI. I've spent some weeks trying to integrate my custom LCD driver with Zephyr's display driver framework.

The driver works fine when I add it to the Zephyr tree (zephyrproject/zephyr/drivers/display/). However, I encounter issues when I try to place it in my own repository and compile it as a Zephyr module; I've tried multiple approaches without success (@gmarull, @jaredwolff). The build system seems to not compile my custom driver, resulting in the exclusion of the display driver from the build and breaking all references to the device instance, which ultimately causes the compilation to fail.

Here is the structure of my project repository:

**my_app_repo/**
├── CMakeLists.txt
├── Kconfig
├── prj.conf
├── west.yml
├── **boards/**
│     ├── my_board.overlay
│     └── **arm/**
│             ├── my_board/
│             ├── board.cmake
│             ├── Kconfig.board
│             ├── Kconfig.defconfig
│             ├── my_board_defconfig
│             ├── my_board.dts
│             └── my_board.yaml
├── **drivers/**
│     ├── CMakeLists.txt
│     ├── Kconfig
│     └── **display/**
│             ├── Kconfig
│             ├── Kconfig.my_lcd
│             ├── CMakeLists.txt
│             ├── my_lcd_driver.c
│             ├── my_lcd_driver.h
│             └── **zephyr/**
│                     └── module.yml
└── **dts/**
        └── **bindings/**
                └── **display/**
                        └── my,compatible.yaml
└── **src/**
       └── main.c

This roadblock has prevented me from demonstrating Zephyr as a viable option for adoption in my company. While the macros related to my driver are generated in autoconf.h, the source files are being ignored during the build.

I would appreciate any guidance or suggestions to resolve this issue.

**Comment by DerekSnell** [comment]
Hi @saacifuentesmu-titoma ,
Are you aware of the Zephyr [example-application repo](https://github.com/zephyrproject-rtos/example-application/tree/main)?  It is meant as a reference for an application repo with out-of-tree resources.  It includes an example of an out-of-tree driver [examplesensor](https://github.com/zephyrproject-rtos/example-application/tree/main/drivers/sensor/examplesensor).  You can setup your driver like this reference.

Best regards

**Comment by saacifuentesmu-titoma** [comment, marked as answer]
Hi @DerekSnell, Just found the [zephyrproject-rtos/example-application](https://github.com/zephyrproject-rtos/example-application) repo some weeks ago.

In summary, I had to follow the next steps:
- Create my repo online.
- Add the west.yml file including the modules.
- Init and Update the workspace locally with west.
- Implement my out of tree project and drivers inside that workspace.

Thank you for the repsonse.
